docs/dev/design/file-tree-diff.rst

```diff
@@ -0,0 +1,234 @@
+  We can store that information as well, or just calculate it on the fly.
+- The list of files are objects, so we can store additional information in the future.
+- When a file has been modified, we also store the number of lines that changed.
+  We could also show this for files that were added or removed.
+- Instead of using the slugs on storage (if we decide to use S3), we could use the IDs, that will prevent the need to update the data if the slugs change.
+  But if the slugs change, the builds will probably be different, so we will need to update the data anyway.
+- If a project or version is deleted (or deactivated), we should delete the diff as well.
+
+We could store the changed files sorted by the number of changes, or make that an option in the API,
+or just let the client sort the files as they see fit.
+
+A combination of using the DB and S3 would be really useful,
+as the model will serve as the lock for the API,
+and will be easier to keep track of which build was used to generate the diff
+without having to download the diff file from S3.
+
+API
+---
+
+This operation is expensive, so it won't be available to unauthenticated users.
+And a diff can only be done between versions of the same project that the user has access to.
+
+The endpoint will be:
+
+   GET /api/v3/projects/{project_slug}/versions/{version_slug}/diff/{other_version_slug}/
+
+And the response will be:
+
+.. code:: json
+
+   {
+       "version_a": {"id": 1, "build": {"id": 1}},
+       "version_b": {"id": 2, "build": {"id": 2}},
+       "diff": {
+           "added": [{"file": "new.txt"}],
+           "removed": [{"file": "deleted.txt"}],
+           "modified": [{"file": "changed.txt", "lines": {"added": 1, "removed": 1}}]
+       }
+   }
+
+The version and build can be the full objects, or just the IDs and slugs.
+
+We will generate a lock on this request, to avoid multiple calls to the API for the same versions.
+We can reply with a ``202 Accepted`` if the diff is being calculated in another request.
+
+Integrations
+------------
+
+You may be thinking that once we have an API, it will be just a matter of calling that API from a GitHub action. Wrong!
+
+Doing the API call is easy, but knowing *when* to call it is hard.
+We need to call the API after the build has finished successfully,
+or we will be comparing the files of an incomplete or stale build.
+
+Luckily, we have a webhook that tells us when a build has finished successfully.
+But, we don't want users to have to implement the integration by themselves.
+
+We could:
+
+- Use this as an opportunity to explore using GitHub Apps.
+- Request additional permissions in our existing OAuth2 integration (``project`` scope). Probably not a good idea.
+- Expose this feature in the dashboard for now, and use our GitHub action to simply link to the dashboard.
+  Maybe don't even expose the API to the public, just use it internally.
```

Comment:
I like having this feature in the dashboard, but the "link me to relevant files changed in this PR" feature would be great for authors as well, which may not have access to the RTD dashboard.

Reply:
I think we initially expose this feature in the Addons PR Build Warning message, which will only be shown on completed builds.

Reply:
Yeah, _that_, but also is something we have 100% control over and we can play with it before going _live_ and let users to use it.